code/modules/projectiles/guns/energy/modular.dm

```diff
@@ -0,0 +1,294 @@
+		if(prob(66))
+			capacitor.small_fail(user)
+		else
+			capacitor.medium_fail(user)
+		qdel(capacitor)
+		capacitor = null
+
+	if(focusing_lens.reliability - focusing_lens.condition <= 0)
+		qdel(focusing_lens)
+		focusing_lens = null
+
+	if(!focusing_lens || !capacitor || !modulator)
+		disassemble()
+		return
+
+	reliability = (capacitor.reliability - capacitor.condition) + (focusing_lens.reliability - focusing_lens.condition)
+
+	if(modulator.projectile)
+		projectile_type = modulator.projectile
+
+	fire_delay = capacitor.fire_delay
+	max_shots = capacitor.shots
+	power_supply.maxcharge = max_shots*charge_cost
+	dispersion = focusing_lens.dispersion
+	accuracy = focusing_lens.accuracy
+	burst += focusing_lens.burst
+
+	if(gun_mods.len)
+		handle_mod()
+
+	fire_delay_wielded = min(0,(fire_delay - fire_delay*3))
+	accuracy_wielded = accuracy + accuracy/4
+	scoped_accuracy = accuracy_wielded + accuracy/4
+	max_shots = max_shots * burst
+	w_class = gun_type
+	reliability = max(reliability, 1)
+
+/obj/item/weapon/gun/energy/laser/prototype/proc/handle_mod()
+	for(var/obj/item/laser_components/modifier/modifier in gun_mods)
+		switch(modifier.mod_type)
+			if(MOD_SILENCE)
+				silenced = 1
+			if(MOD_NUCLEAR_CHARGE)
+				self_recharge = 1
+				criticality *= 2
+		fire_delay *= modifier.fire_delay
+		reliability += modifier.reliability
+		burst += modifier.burst
+		burst_delay += modifier.burst_delay
+		max_shots *= modifier.shots
+		force += modifier.gun_force
+		chargetime += modifier.chargetime*10
+		accuracy += modifier.accuracy
+		criticality *= modifier.criticality
+		if(modifier.scope_name)
+			zoomdevicename = modifier.scope_name
+
+/obj/item/weapon/gun/energy/laser/prototype/consume_next_projectile()
+	if(!power_supply)
+		return null
+	if(!ispath(projectile_type))
+		return null
+	if(!power_supply.checked_use(charge_cost))
+		return null
+	if (self_recharge)
+		addtimer(CALLBACK(src, .proc/try_recharge), recharge_time * 2 SECONDS, TIMER_UNIQUE)
+	var/obj/item/projectile/beam/A = new projectile_type(src)
+	A.damage = capacitor.damage
+	var/damage_coeff = 1
+	for(var/obj/item/laser_components/modifier/modifier in gun_mods)
+		damage_coeff *= modifier.damage
+	if(burst > 1)
+		A.damage = A.damage/(burst - 1)
+	damage_coeff *= modulator.damage
+	A.damage *= damage_coeff
+	A.damage = min(A.damage, 60) //let's not get too ridiculous here
+	for(var/obj/item/laser_components/modifier/modifier in gun_mods)
+		if(prob((gun_mods.len * 10 * damage_coeff)/(max(1,(burst - 1)))))
+			capacitor.degrade(modifier.malus)
+		if(prob((gun_mods.len * 10 * damage_coeff)/(max(1,(burst - 1)))))
+			focusing_lens.degrade(modifier.malus)
+		if(prob((33 + capacitor.damage)/(max(1,(burst - 1)))))
+			modifier.degrade(1)
+	if((prob(A.damage)/burst))
+		if(prob(A.damage/2))
+			medium_fail(user)
+		else
+			small_fail(user)
+
+	updatetype()
+	return A
+
+/obj/item/weapon/gun/energy/laser/prototype/proc/disassemble()
+	var/turf/T = get_turf(src)
```

Comment:
This will generate issues, once again.

If a mob is inside a locker and disassembles the gun, it'll move the pieces into the turf, and not the locker. Perhaps pass `user` as an argument and do this:
```DM
/obj/item/weapon/gun/energy/laser/prototype/proc/disassemble(mob/user)
	var/atom/A
	if (user && user.loc)
		A = user.loc
	else
		A = get_turf(src)
```

This'll move it inside the user's location first, and then onto the turf second.

Reply:
I think there's a proc for handling this automatically?
`drop_into()` or something.